ocaml/xapi/xapi_vm_lifecycle.ml

```diff
@@ -367,194 +365,207 @@ let maybe_get_guest_metrics ~__context ~ref =
+
+     (* Always check the power state constraint of the operation first *)
+     let current_error = check current_error (fun () ->
+         if not (is_allowed_sequentially ~__context ~vmr ~power_state ~op)
+         then report_power_state_error ~__context ~vmr ~power_state ~op ~ref_str
+         else None) in
+
+     (* if other operations are in progress, check that the new operation is allowed concurrently with them. *)
+     let current_error = check current_error (fun () ->
+         let current_ops = vmr.Db_actions.vM_current_operations in
+         if List.length current_ops <> 0 && not (is_allowed_concurrently ~op ~current_ops)
+         then report_concurrent_operations_error ~current_ops ~ref_str
+         else None) in
+
+     (* if the VM is a template, check the template behavior exceptions. *)
+     let current_error = check current_error (fun () ->
+         if is_template && not is_snapshot
+         then check_template ~vmr ~op ~ref_str
+         else None) in
+
+     (* if the VM is a snapshot, check the snapshot behavior exceptions. *)
+     let current_error = check current_error (fun () ->
+         if is_snapshot
+         then check_snapshot ~vmr ~op ~ref_str
+         else None) in
+
+     (* if the VM is neither a template nor a snapshot, do not allow provision and revert. *)
+     let current_error = check current_error (fun () ->
+         if op = `provision && (not is_template)
+         then Some (Api_errors.only_provision_template, [])
+         else None) in
+
+     let current_error = check current_error (fun () ->
+         if op = `revert && (not is_snapshot)
+         then Some (Api_errors.only_revert_snapshot, [])
+         else None) in
+
+     (* Some ops must be blocked if VM is not mobile *)
+     let current_error = check current_error (fun () ->
+         match op with
+         | `suspend
+         | `checkpoint
+         | `pool_migrate
+         | `migrate_send
+           when not (is_mobile ~__context strict metrics platformdata) ->
+           Some (Api_errors.vm_is_immobile, [ref_str])
+         | _ -> None
+       ) in
+
+     let current_error =
+       check current_error (fun () ->
+           match op with
+           | `changing_dynamic_range
+             when nested_virt ~__context metrics platformdata && strict ->
+             Some (Api_errors.vm_is_using_nested_virt, [ref_str])
+           | _ -> None
+         ) in
+
+
+     (* Check if the VM is a control domain (eg domain 0).            *)
+     (* FIXME: Instead of special-casing for the control domain here, *)
+     (* make use of the Helpers.ballooning_enabled_for_vm function.   *)
+     let current_error = check current_error (fun () ->
+         if (op = `changing_VCPUs || op = `destroy) && Helpers.is_domain_zero ~__context !!vm_ref
+         then Some (Api_errors.operation_not_allowed, ["This operation is not allowed on dom0"])
+         else if vmr.Db_actions.vM_is_control_domain
+              && op <> `data_source_op
```

Comment:
Using pattern matching here instead of spearate manual checks would be optimized by the compiler, but I am not sure it's worth it.